icicle_v3/backend/cpu/src/curve/cpu_msm.hpp

```diff
@@ -0,0 +1,667 @@
+ */
+eIcicleError not_supported(const MSMConfig& conf)
+{
+  if (conf.is_async) { return eIcicleError::INVALID_DEVICE; }
+  // There is only host (CPU) therefore the following configs are not supported:
+  if (conf.are_scalars_on_device | conf.are_points_on_device | conf.are_results_on_device) {
+    return eIcicleError::INVALID_DEVICE;
+  }
+  return eIcicleError::SUCCESS;
+}
+
+/**
+ * @brief Super function that handles the Msm class to calculate a MSM.
+ * @param device - Icicle API parameter stating the device being ran on. In this case - CPU.
+ * @param scalars - Input scalars for MSM.
+ * @param bases - EC P input, affine representation.
+ * @param msm_size - Size of the above arrays, as they are given as Pers.
+ * @param config - configuration containing parameters for the MSM.
+ * @param results - Per to P array in which to store the results. NOTE: the user is expected to preallocate the
+ *                  results array.
+ */
+template <typename A, typename P>
+eIcicleError cpu_msm(
+  const Device& device, const scalar_t* scalars, const A* bases, int msm_size, const MSMConfig& config, P* results)
+{
+  Msm<A, P>* msm = new Msm<A, P>(config);
+
+  if (not_supported(config) != eIcicleError::SUCCESS) return not_supported(config);
+
+  const unsigned int c = config.c;
+  const unsigned int precompute_factor = config.precompute_factor;
+  const int num_bms = ((scalar_t::NBITS - 1) / (precompute_factor * c)) + 1;
+
+  if (config.ext->get<int>("n_threads") <= 0) { return eIcicleError::INVALID_ARGUMENT; }
+
+  for (int i = 0; i < config.batch_size; i++) {
+    msm->run_msm(&scalars[msm_size * i], bases, msm_size, i, &results[i]);
+  }
+  delete msm;
+  return eIcicleError::SUCCESS;
+}
+
+/**
+ * @brief Function to precompute Ps multiplications - trading memory for MSM performance.
+ * @param device - Icicle API parameter stating the device being ran on. In this case - CPU.
+ * @param input_bases - Ps to precompute.
+ * @param nof_bases - Size of the above array, as it is given as a Per.
+ * @param config - configuration containing parameters for the MSM. In this case, the config implicitly determines the
+ *                 multiplication factor(s) of the input bases.
+ * @param output_bases - Per to P array in which to store the results. NOTE: the user is expected to
+ *                       preallocate the results array.
+ */
+template <typename A, typename P>
+eIcicleError cpu_msm_precompute_bases(
+  const Device& device,
+  const A* input_bases,
+  int nof_bases,
+  const MSMConfig& config,
+  A* output_bases) // Pre assigned?
+{
+  int precompute_factor = config.precompute_factor;
+  bool is_mont = config.are_points_montgomery_form;
+  // bool is_mont=false;
+  const unsigned int c = config.c;
+  const unsigned int num_bms_no_precomp = (scalar_t::NBITS - 1) / c + 1;
+  const unsigned int shift = c * ((num_bms_no_precomp - 1) / precompute_factor + 1);
+  for (int i = 0; i < nof_bases; i++) {
```

Comment:
is that not going to be parallel? Do you know approximately how long it takes for typical cases? It's done once but still if slow can be an issue.

Reply:
Currently not
It can be added in the future (Maybe with the preprocess for phase 1 mentioned above)